xclim/indices/_threshold.py

```diff
@@ -1311,10 +1316,17 @@ def first_day_temperature_above(
     )
 
 
-@declare_units(prsn="[precipitation]", thresh="[precipitation]")
+@declare_units(
+    prsn="[precipitation]",
+    thresh="[precipitation]",
+    snr="[mass]/[volume]",
+    const="[mass]/[volume]",
+)
 def first_snowfall(
     prsn: xarray.DataArray,
     thresh: Quantified = "0.5 mm/day",
+    snr: xarray.DataArray | None = None,
+    const: Quantified = "312 kg m-3",
     freq: str = "AS-JUL",
 ) -> xarray.DataArray:
     r"""First day with solid precipitation above a threshold.
```

Comment:
@Zeitsperre Could you give your insight on this. I realize that this index (and the next one) differs from how `pr` indices usually work, and I want to know if we want to allow this extra layer of complexity.

Indices with `pr` usually simply keep the units of `pr` and simply adapt the threshold if needed. This is OK since the conversion is always uniform, same density 1000 kg m-3 everywhere regardless. The case of `prsn` with possibly space-dependent density (`prsn`) is more complex. 

Should we:
1. Allow the conversion of `prsn` inside the index as things are now? 
2. Only let the index convert `thresh` with a constant conversion factor if needed as in `pr` indices. In this case, one would need to convert `prsn` if [length]/[time] units are desired.

Maybe the confusion also comes from the fact that [precipitation] can mean both [mass]/[area]/[time] and [length]/[time]. In case of actual `pr`, this is not so ambiguous, since the conversion factor is uniform. But for `prsn`, things are more complex. We were discussing how `prsn` should be reserved for [mass]/[area]/[time], and there should be another variable name when units are [length]/[time].

Reply:
I can see how quickly this case is becoming complicated; I think we need to agree on some approaches for handling precipitation. The CF Conventions are pretty specific on what counts as Precipitation Flux (`pr`; [mass]/[area]/[time]) and Precipitation Amount (no official acronym; [mass]/[area]).

I don't like the idea of hiding conversions within an indice, especially if this conversion step might be used in multiple indices. We don't currently log these conversion steps into the metadata history either, making matters worse. I do think we should be able to support amounts and fluxes within an indicator/index, but how we handle the conversion needs to be clearly communicated (via warnings?) to the user.

The current behaviour to only use a rule-of-thumb liquid water equivalent conversion only if `snr` is not provided feels like a good approach, so I wouldn't change that. 

The more I think about it, the more I agree that we need to define two variables for snow amount and flux. Or at the very least have some code to handle the case of prsn as an amount.